File: ios/PouchProv/Sources/PouchProvCore/ProvSession.swift
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// Seam between the RPC engine and the pouch framing/crypto. The plaintext
/// implementation frames encryption-none pouches; a saead implementation can be
/// dropped in later without reshaping ``ProvSession``.
public protocol SessionCrypto: Sendable {
    /// Build the request (downlink) pouch bytes carrying `entries`.
    func buildRequest(entries: [Entry], deviceId: String, blockSize: Int) throws -> Data

    /// Parse a response (uplink) pouch into its header and entries.
    func parseResponse(_ data: Data) throws -> (header: PouchHeader, entries: [Entry])
}

/// Plaintext (encryption-none) framing — parity with the CLI's live path.
public struct PlaintextCrypto: SessionCrypto {
    public init() {}

    public func buildRequest(entries: [Entry], deviceId: String, blockSize: Int) throws -> Data {
        try Pouch.buildEntryPouch(
            deviceId: deviceId.isEmpty ? "?" : deviceId,
            entries: entries,
            blockSize: blockSize
        )
    }

    public func parseResponse(_ data: Data) throws -> (header: PouchHeader, entries: [Entry]) {
        try Pouch.parsePouch(data)
    }
}

public struct SessionError: Error, CustomStringConvertible {
    public let message: String

    public init(_ message: String) {
        self.message = message
    }

    public var description: String { message }
}

/// Lockstep request/response engine. One request pouch at a time.
///
/// Implements the client-driven RPC cycle (docs/protocol.md): request = subscribe
/// downlink -> SAR-send one pouch -> unsubscribe; response = subscribe uplink ->
/// SAR-receive one pouch -> unsubscribe. An empty response pouch means "not
/// ready, retry". A faithful port of the Kotlin/Python `ProvSession`.
public final class ProvSession: @unchecked Sendable {
    // ATT MTU 247 => 244-byte payloads on notification / write.
    public static let defaultMaxlen = 244

    private let transport: any ProvTransport
    public var deviceId: String
    private let maxlen: Int
    private let crypto: any SessionCrypto

    public var blockSize: Int = Pouch.defaultBlockSize

    private let lock = AsyncMutex()

    public init(
        transport: any ProvTransport,
        deviceId: String = "",
        maxlen: Int = ProvSession.defaultMaxlen,
        crypto: any SessionCrypto = PlaintextCrypto()
    ) {
        self.transport = transport
        self.deviceId = deviceId
        self.maxlen = maxlen
        self.crypto = crypto
    }

    /// Send request entries in one pouch, return the response entries.
    public func requestEntries(
        _ entries: [Entry],
        timeoutMs: Int64 = 15_000,
        pollAttempts: Int = 5
    ) async throws -> [Entry] {
        try await lock.withLock {
            try await sendPouch(entries, timeoutMs: timeoutMs)

            var backoff: Int64 = 200
            for _ in 0..<pollAttempts {
                let (header, rsp) = try await receivePouch(timeoutMs: timeoutMs)
                deviceId = header.deviceId ?? deviceId
                if !rsp.isEmpty { return rsp }
                // Empty pouch: responses were not ready; back off and re-poll.
                try await Task.sleep(nanoseconds: UInt64(backoff) * 1_000_000)
                backoff = min(backoff * 2, 2_000)
            }
            throw SessionError("no response after retries")
        }
    }

    /// Single-entry convenience: request `message` on `path`, return the response bytes.
    public func request(path: String, message: Data, timeoutMs: Int64 = 15_000) async throws -> Data {
        let entries = try await requestEntries(
            [Entry(path: path, contentType: Pouch.contentTypeCbor, data: message)],
            timeoutMs: timeoutMs
        )
        guard let rsp = entries.first(where: { $0.path == path })?.data else {
            throw SessionError("no response entry for \(path)")
        }
        return rsp
    }

    private func sendPouch(_ entries: [Entry], timeoutMs: Int64) async throws {
        let data = try crypto.buildRequest(entries: entries, deviceId: deviceId, blockSize: blockSize)
        let channel = transport.downlink
        let sender = SarSender(write: { try await channel.write($0) }, maxlen: maxlen)
        try await channel.subscribe { sender.feed($0) }
        do {
            try await sender.send(data, timeoutMs: timeoutMs)
        } catch {
            await channel.unsubscribe()
            throw error
        }
        await channel.unsubscribe()
    }

    private func receivePouch(timeoutMs: Int64) async throws -> (header: PouchHeader, entries: [Entry]) {
        let channel = transport.uplink
        let receiver = SarReceiver(write: { try await channel.write($0) })
        try await channel.subscribe { receiver.feed($0) }
        let data: Data
        do {
            data = try await receiver.receive(timeoutMs: timeoutMs)
        } catch {
            await channel.unsubscribe()
            throw error
        }
        await channel.unsubscribe()
        return try crypto.parseResponse(data)
    }
}

File: android/pouchprov-core/src/main/kotlin/io/golioth/pouchprov/session/ProvSession.kt
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

package io.golioth.pouchprov.session

import io.golioth.pouchprov.pouch.Entry
import io.golioth.pouchprov.pouch.Pouch
import io.golioth.pouchprov.pouch.PouchHeader
import io.golioth.pouchprov.saead.Saead
import io.golioth.pouchprov.saead.SaeadSession
import io.golioth.pouchprov.sar.SarReceiver
import io.golioth.pouchprov.sar.SarSender
import io.golioth.pouchprov.transport.Transport
import java.security.SecureRandom
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Seam between the RPC engine and the pouch framing/crypto. The plaintext
 * implementation frames encryption-none pouches; a saead implementation can be
 * dropped in later without reshaping [ProvSession] (see the plan's crypto seam).
 */
interface SessionCrypto {
    /** Build the request (downlink) pouch bytes carrying [entries]. */
    fun buildRequest(entries: List<Entry>, deviceId: String, blockSize: Int): ByteArray

    /** Parse a response (uplink) pouch into its header and entries. */
    fun parseResponse(data: ByteArray): Pair<PouchHeader, List<Entry>>
}

/** Plaintext (encryption-none) framing — parity with the CLI's live path. */
object PlaintextCrypto : SessionCrypto {
    override fun buildRequest(entries: List<Entry>, deviceId: String, blockSize: Int): ByteArray =
        Pouch.buildEntryPouch(deviceId.ifEmpty { "?" }, entries, blockSize)

    override fun parseResponse(data: ByteArray): Pair<PouchHeader, List<Entry>> = Pouch.parsePouch(data)
}

/**
 * Encrypted (saead) framing over a [SaeadSession]. Each request pouch starts a
 * fresh random downlink session; the uplink session is adopted from the device's
 * response header. An empty (header-only) uplink pouch means "not ready".
 */
class SaeadCrypto(
    private val saead: SaeadSession,
    private val sessionIdSource: () -> ByteArray = {
        ByteArray(Saead.SESSION_ID_LEN).also { SecureRandom().nextBytes(it) }
    },
) : SessionCrypto {
    override fun buildRequest(entries: List<Entry>, deviceId: String, blockSize: Int): ByteArray =
        Saead.buildDownlinkPouch(saead, sessionIdSource(), entries, blockSize = blockSize)

    override fun parseResponse(data: ByteArray): Pair<PouchHeader, List<Entry>> {
        val (header, consumed) = PouchHeader.decode(data)
        if (header.encryption == Pouch.ENCRYPTION_SAEAD) {
            // A header-only uplink pouch still parses to no entries ("not ready").
            return if (data.size > consumed) header to Saead.parseUplinkPouch(saead, data) else header to emptyList()
        }
        return Pouch.parsePouch(data)
    }
}

class SessionError(message: String) : Exception(message)

/**
 * Lockstep request/response engine. One request pouch at a time.
 *
 * Implements the client-driven RPC cycle (docs/protocol.md): request = subscribe
 * downlink -> SAR-send one pouch -> unsubscribe; response = subscribe uplink ->
 * SAR-receive one pouch -> unsubscribe. An empty response pouch means "not
 * ready, retry". A faithful port of the Python `session.ProvSession`.
 */
class ProvSession(
    private val transport: Transport,
    var deviceId: String = "",
    private val maxlen: Int = DEFAULT_MAXLEN,
    private val crypto: SessionCrypto = PlaintextCrypto,
) {
    var blockSize: Int = Pouch.DEFAULT_BLOCK_SIZE

    private val lock = Mutex()

    /** Send request entries in one pouch, return the response entries. */
    suspend fun requestEntries(
        entries: List<Entry>,
        timeoutMs: Long = 15_000,
        pollAttempts: Int = 5,
    ): List<Entry> = lock.withLock {
        sendPouch(entries, timeoutMs)

        var backoff = 200L
        repeat(pollAttempts) {
            val (header, rsp) = receivePouch(timeoutMs)
            deviceId = header.deviceId ?: deviceId
            if (rsp.isNotEmpty()) return@withLock rsp
            // Empty pouch: responses were not ready; back off and re-poll.
            delay(backoff)
            backoff = minOf(backoff * 2, 2_000L)
        }
        throw SessionError("no response after retries")
    }

    /** Single-entry convenience: request [msg] on [path], return the response bytes. */
    suspend fun request(path: String, msg: ByteArray, timeoutMs: Long = 15_000): ByteArray {
        val entries = requestEntries(listOf(Entry(path, Pouch.CONTENT_TYPE_CBOR, msg)), timeoutMs)
        return entries.firstOrNull { it.path == path }?.data
            ?: throw SessionError("no response entry for $path")
    }

    private suspend fun sendPouch(entries: List<Entry>, timeoutMs: Long) {
        val data = crypto.buildRequest(entries, deviceId, blockSize)
        val ch = transport.downlink
        val sender = SarSender(ch::write, maxlen)
        ch.subscribe(sender::feed)
        try {
            sender.send(data, timeoutMs)
        } finally {
            ch.unsubscribe()
        }
    }

    private suspend fun receivePouch(timeoutMs: Long): Pair<PouchHeader, List<Entry>> {
        val ch = transport.uplink
        val receiver = SarReceiver(ch::write)
        ch.subscribe(receiver::feed)
        val data = try {
            receiver.receive(timeoutMs)
        } finally {
            ch.unsubscribe()
        }
        return crypto.parseResponse(data)
    }

    companion object {
        // ATT MTU 247 => 244-byte payloads on notification / write.
        const val DEFAULT_MAXLEN = 244
    }
}
